Defcon - SPAWN_DEFENSE_ON_ATTACK

BringerOD opened the pull request:
This is up to date and ready to merge.

I changed this to a Boolean setting that when set,  attempts to store enough energy for a fight, then will spawn troops,  when attacked.

You do not need a flag.  If you have a flag for defense, this is additive.

Let me know what you think.

cyberblast commented:
Hey, sounds great! :+1: 
Unfortunately I got to go for now. :( 
I will have a look at it later.

BringerOD commented:
Special note here.

I stop producing creeps of various types to ensure there is enough energy, once there is, then those troops can get spawned. 

Also,  I added some logic not to spawn a worker unless there is some work to do, and changed the upgrader count to 2. We can remove this if you want, but its working nicely for me.

mikiones commented on creep.setup.worker.js:
```diff
@@ -7,6 +7,9 @@ setup.minEnergyAvailable = function(spawn){
 setup.maxCount = function(spawn){
+    if (spawn.room.situation.invasion) return 0;  // Do not create in the middle of a fight.  We should be building fighters if we have energy left.
+    if (setup.ShouldWeConserveForDefense(spawn)) return 0;
+    if (spawn.room.constructionSites.length == 0)  return 0;
```
check if we have tower and if nothing to retire first

BringerOD replied:
Not sure you want to spawn an entire worker for just a few repairable.  Maybe just urgent repairables.

mikiones commented on creep.setup.upgrader.js:
```diff
@@ -7,7 +7,13 @@ setup.minEnergyAvailable = function(spawn){
+    let numberToBuild = 1;
+    if (spawn.room.situation.invasion) return 0;  // Do not create in the middle of a fight
+    if (setup.ShouldWeConserveForDefense(spawn)) return 0;
+    if (spawn.room.constructionSites.length == 0)  numberToBuild +=1;
```
would be nice if we check how full the container is if > 90% spawn additional upgraders

BringerOD commented:
Good catch!  Agreed when storage is > x% spawn more upgraders.

BringerOD commented:
Is there a way to accept your change?

cyberblast commented:
Why do you want to have more upgraders? 
More creeps mean less cpu & less space. 
We should stick to the minimalistic approach and instead adjust creep size... (at least until > 50 body parts required)...

cyberblast commented:
Can we please remove the worker/upgrader adjustment for the moment? To have the pure defcon solution as a first step. I think we all agree on this part (?). And I would like to merge that. 
Of course we may discuss further steps, but I don't want to mix that...

BringerOD commented:
sure I will remove.

BringerOD commented:
Done.

cyberblast commented:
Creep size is not only limited to 50 parts, its also the max available energy all your extensions can hold to invest in a creep at once. 
So .. there indeed is the requirement to make more upgraders possible. 
But at first the upgrader should grow. If then the storage level ist still increasing, an additional upgrader will be allowed.

cyberblast commented:
I'm unsure regarding the worker. 
The worker also supports with feeding, fueling etc. 

On the other hand there is nothing ONLY the worker can do, except building. And there are idle times for the worker. 

I think we should add an issue for it to implement at a later time. I dont think its that expensive keeping the worker for the moment. What do you think?
